compose/ui/ui/src/webCommonW3C/kotlin/androidx/compose/ui/platform/DomInputStrategy.kt

```diff
@@ -0,0 +1,174 @@
+package androidx.compose.ui.platform
+
+import androidx.compose.ui.text.input.CommitTextCommand
+import androidx.compose.ui.text.input.DeleteSurroundingTextInCodePointsCommand
+import androidx.compose.ui.text.input.EditCommand
+import androidx.compose.ui.text.input.SetComposingTextCommand
+import androidx.compose.ui.text.input.TextFieldValue
+import org.w3c.dom.HTMLTextAreaElement
+import org.w3c.dom.events.CompositionEvent
+import org.w3c.dom.events.Event
+import org.w3c.dom.events.KeyboardEvent
+
+internal interface DomInputStrategy {
+    val htmlInput: HTMLTextAreaElement
+    fun updateState(textFieldValue: TextFieldValue)
+}
+
+internal class CommonDomInputStrategy(
+    override val htmlInput: HTMLTextAreaElement,
+    private val composeSender: ComposeCommandCommunicator,
+) : DomInputStrategy {
+
+    private var lastActualCompositionTimestamp: Int = 0
+    private var editState: EditState = EditState.Default
+
+
+    init {
+        initEvents()
+    }
+
+    override fun updateState(textFieldValue: TextFieldValue) {
+        if (editState != EditState.WaitingComposeActivity) return
+
+        htmlInput.value = textFieldValue.text
+        htmlInput.setSelectionRange(textFieldValue.selection.start, textFieldValue.selection.end)
+
+        editState = EditState.Default
+    }
+
+    private fun initEvents() {
+        var lastKeyboardEventIsDown = false
+
+        htmlInput.addEventListener("blur", {evt ->
+            // both accent dialogue and composition dialogue are lost when we switch windows
+            // but can be restored later on if we are back
+            editState = EditState.Default
+        })
+
+        htmlInput.addEventListener("keydown", {evt ->
+            evt as KeyboardEvent
+            lastKeyboardEventIsDown = evt.key != "Dead" && evt.key != "Unidentified"
+
+            if (editState is EditState.AccentDialogue) {
+                evt.preventDefault()
+                return@addEventListener
+            }
+
+            if (evt.repeat) {
+                editState = EditState.AccentDialogue
+                evt.preventDefault()
+                return@addEventListener
+            }
+
+            if (evt.isComposing) {
+                editState = EditState.CompositeDialogue
+            }
+
+            if (editState is EditState.CompositeDialogue) {
+                evt.preventDefault()
+                return@addEventListener
+            }
+
+            if (evt.repeat) {
+                editState = EditState.AccentDialogue
+            }
+
+            if (editState is EditState.AccentDialogue) {
```

Comment:
This condition is already checked above (line 54).

Reply:
We need both - we can end up here because we've just end up in this state